shell/platform/windows/client_wrapper/flutter_window_controller.cc

```diff
@@ -0,0 +1,407 @@
+    return window->MessageHandler(hwnd, message, wparam, lparam);
+  }
+  return DefWindowProc(hwnd, message, wparam, lparam);
+}
+
+void FlutterWindowController::SendOnWindowCreated(
+    FlutterViewId view_id,
+    std::optional<FlutterViewId> parent_view_id) const {
+  if (channel_) {
+    channel_->InvokeMethod(
+        "onWindowCreated",
+        std::make_unique<EncodableValue>(EncodableMap{
+            {EncodableValue("viewId"), EncodableValue(view_id)},
+            {EncodableValue("parentViewId"),
+             parent_view_id ? EncodableValue(parent_view_id.value())
+                            : EncodableValue()}}));
+  }
+}
+
+void FlutterWindowController::SendOnWindowDestroyed(
+    FlutterViewId view_id) const {
+  if (channel_) {
+    channel_->InvokeMethod(
+        "onWindowDestroyed",
+        std::make_unique<EncodableValue>(EncodableMap{
+            {EncodableValue("viewId"), EncodableValue(view_id)},
+        }));
+  }
+}
+
+void FlutterWindowController::SendOnWindowChanged(FlutterViewId view_id) const {
+  if (channel_) {
+    auto const size{GetWindowSize(view_id)};
+    channel_->InvokeMethod(
+        "onWindowChanged",
+        std::make_unique<EncodableValue>(EncodableMap{
+            {EncodableValue("viewId"), EncodableValue(view_id)},
+            {EncodableValue("size"),
+             EncodableValue(EncodableList{EncodableValue(size.width),
+                                          EncodableValue(size.height)})},
+            {EncodableValue("relativePosition"), EncodableValue()},  // TODO
+            {EncodableValue("isMoving"), EncodableValue()}}));       // TODO
+  }
+}
+
+void FlutterWindowController::HandleCreateWindow(WindowArchetype archetype,
+                                                 MethodCall<> const& call,
+                                                 MethodResult<>& result) {
+  auto const* const arguments{call.arguments()};
+  auto const* const map{std::get_if<EncodableMap>(arguments)};
+  if (!map) {
+    result.Error(kErrorCodeInvalidValue, "Method call argument is not a map.");
+    return;
+  }
+
+  std::wstring const title{ArchetypeToWideString(archetype)};
+
+  auto const size_list{
+      GetListValuesForKeyOrSendError<int, 2>("size", map, result)};
+  if (!size_list) {
+    return;
+  }
+  if (size_list->at(0) < 0 || size_list->at(1) < 0) {
+    result.Error(kErrorCodeInvalidValue,
+                 "Values for 'size' key (" + std::to_string(size_list->at(0)) +
+                     ", " + std::to_string(size_list->at(1)) +
+                     ") must be nonnegative.");
+    return;
+  }
+
+  if (auto const data_opt{CreateFlutterWindow(
+          title, {.width = size_list->at(0), .height = size_list->at(1)},
+          archetype)}) {
+    auto const& data{data_opt.value()};
+    result.Success(EncodableValue(EncodableMap{
+        {EncodableValue("viewId"), EncodableValue(data.view_id)},
+        {EncodableValue("archetype"),
```

Comment:
Enumerated values in other channels are normally passed as strings, i.e. "Archetype.regular"

Reply:
Good to know, thanks. I'll change it.